[BugFix] Fix hive catalog can't read hive table with textfile lzo format bug (part 2: add LzoStreamCompression)

sfwang218 opened the pull request:
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #19908 

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
Hive catalog can't read hive table with textfile lzo format

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

choury committed:
[BugFix] fix parse error of QUALIFY with alias

Signed-off-by: choury <chouryzhou@tencent.com>

sfwang218 committed:
Add LzoStreamCompression to support text lzo file

Signed-off-by: wangshifa <wangshifa@58.com>

silverbullet233 committed:
change default dop back to half of cpu cores (#19966)

Signed-off-by: silverbullet233 <3675229+silverbullet233@users.noreply.github.com>

xiaoyong-z committed:
[BugFix] Incorrect query during compaction with delete predicates (#20170)

For duplicate tablet with delete predicate rowset, compaction will
merge serveral rowsets(with delete predicate rowset) into 1 large
rowset(without delete predicate). After compaction, stale rowsets' delete
predications will be removed in tablet_meta(TabletMeta::modify_rs_metas function).

There exists an contention case if a query during compaction with delete predicates.

For example, if table's rowset is [0-2],[3-3], rowset [3,3] is delete rowset
And query and compaction happen in the following order:
(1) a query come before the compaction, and its version is [0,3],
then TabletReader will select rowset [0,2] and [3,3] in prepare().
(2) Compaction happens, rowset [0,2] and [3,3] are merged into rowset [0,3],
rowset [0-3] doesn't contain delete predicate. After that,
delete predicates for [3,3] is removed in _del_pred_array of tablet_meta.
(3) TabletReader executes TabletReader::open and _init_delete_predicates,
because delete predicates for [3,3] has been removed in tablet_meta, then
_init_delete_predicates will return 0 delete predicates, and the query will ignore the
delete predicate of [3,3], this will result in incorrect query result.

Signed-off-by: xyz <a997647204@gmail.com>

HangyuanLiu committed:
Activate_all_roles_on_login is changed to Boolean type and set to global 

Signed-off-by: HangyuanLiu <460660596@qq.com>

HangyuanLiu committed:
Fix no database permission but use catalog.db is successful bug (#20234)

Signed-off-by: HangyuanLiu <460660596@qq.com>

amber-create committed:
[Doc] Update system variables doc (#20261)

Signed-off-by: amber-create <yangyanping@starrocks.com>

sduzh committed:
[BugFix] Fix parse empty array throw NPE (#20226)

Fix #20215

Signed-off-by: sduzh <zhuming9011@qq.com>

sduzh committed:
[Feature] Support query table function directly (#19973)

Now we can query table functions like normal tables:
```
> select * from TABLE(unnest(ARRAY<INT>[1,2])) t(x); // select
+------+
| x    |
+------+
|    1 |
|    2 |
+------+
2 rows in set (0.02 sec)

> select * from t0; // t0 is a normal table
+------+
| c0   |
+------+
|    1 |
+------+
1 row in set (0.10 sec)

> select * from t0, TABLE(unnest(ARRAY<INT>[1, 2])); // cross join
+------+--------+
| c0   | unnest |
+------+--------+
|    1 |      1 |
|    1 |      2 |
+------+--------+
2 rows in set (0.03 sec)

> select * from t0 left join TABLE(unnest(ARRAY<INT>[1, 2])) t1(x) on t0.c0=t1.x; // left join
+------+------+
| c0   | x    |
+------+------+
|    1 |    1 |
+------+------+
1 row in set (0.06 sec)

> select * from t0 right join TABLE(unnest(ARRAY<INT>[1, 2])) t1(x) on t0.c0=t1.x; // right join
+------+------+
| c0   | x    |
+------+------+
|    1 |    1 |
| NULL |    2 |
+------+------+
2 rows in set (0.04 sec)
```

Signed-off-by: sduzh <zhuming9011@qq.com>

kevincai committed:
[Tool][Dockerfile] add ubuntu toolchains dockerfile (#20284)

ABingHuang committed:
[Bugfix] fix rewrite bug after insert new partition data (#20157)

Fix rewrite failure after inserting new partition data. 
The bug reason is that the mv plan is cached, but the plan may change after ingestion with new partitions data.
And the logic of partition predicate calculation for mv rewrite depends on scan node's selected partition id,
which may change after ingestion. So it leads to invalid compensation predicate in mv rewrite,
which caused an invalid rewritten plan. Fix it by:
1. disable partition prune rules during compile mv plan, which will keep partition predicates in mv plan, to avoid the problem of caching plan
2. remove mv partition predicate compensation logic because it is not necessary after step 1. and it also fixed the problem of redundant partition predicates after partition prune of mv.

Signed-off-by: ABingHuang <codekhuang@163.com>

before-Sunrise committed:
[Feature] support agg function percentile_disc (#17628)

Signed-off-by: before-Sunrise <unclejyj@gmail.com>

imay commented on be/CMakeLists.txt:
```diff
@@ -193,6 +195,11 @@ set_target_properties(curl PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib/li
 add_library(lz4 STATIC IMPORTED)
 set_target_properties(lz4 PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib/liblz4.a)
 
+if (${WITH_LZO} STREQUAL "ON")
```
It will be better to use "$ENV{WITH_LZO}" 
Then there is no need to change build.sh and run-be-ut.sh

imay commented on be/CMakeLists.txt:
```diff
@@ -720,6 +727,13 @@ if ("${CMAKE_BUILD_TARGET_ARCH}" STREQUAL "x86" OR "${CMAKE_BUILD_TARGET_ARCH}"
+        lzo
+    )
+endif()
+
```
you need to `set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -DWITH_LZO") ` when ENV{WITH_LZO} is true, then the compiler will define the WITH_LZO macro when compiling the code.

imay commented on be/src/util/compression/stream_compression.cpp:
```diff
@@ -422,6 +426,455 @@ Status ZstandardStreamCompression::decompress(uint8_t* input, size_t input_len,
 
+// lzo
+#if defined(WITH_LZO)
+class LzoStreamCompression : public StreamCompression {
```
You'd better add Unit Tests for this newly added method.
It is easy to reuse the existing tests.

dirtysalt committed:
[BugFix] fix cmake file in block cache submodule (#20291)

xiaoyong-z committed:
Revert "[BugFix] Incorrect query during compaction with delete predicates (#20170)" (#20293)

dirtysalt committed:
[Refactor] clean some be configs rarely modified (#20290)

Signed-off-by: yanz <dirtysalt1987@gmail.com>

dirtysalt committed:
[Enhancement] put `io_task_per_scan_operator` to session variable (#20289)

dirtysalt committed:
fix to avoid scan op is nullptr when memlimit is 0 (#20300)

rickif committed:
[BugFix] Non-existed jsonpath (#19358)

This PR solves some jsonpath issues. Please refer the test case in be/test/exprs/json_functions_test.cpp to know the change.

kevincai committed:
[Tool][Dockerfile] unify dev-env dockerfile (#20309)

wyb committed:
[BugFix] Cancel fragment instance not query when close external scanner (#20264)

* [BugFix] cancel fragment instance not query when close external scanner

If cancel query, other fragment instances (scanners) will be cancelled,
and spark connector will report SocketTimeoutException when get next data from BE.

Query executed by external connectors does not need to report exec state
to FE.

Signed-off-by: wyb <wybb86@gmail.com>

wyb committed:
[BugFix] Storage cache ttl should be 0 when disable storage cache (#20185)

If storage_cache_ttl is not specified, it should be assigned according to whether storage cache is enabled.

Signed-off-by: wyb <wybb86@gmail.com>

sfwang218 committed:
Use ENV replace option

Signed-off-by: wangshifa <wangshifa@58.com>

sfwang218 committed:
Merge pull request #1 from StarRocks/main

rebase
